// @neo4j/graphql/dist/translate/connection/create-connection-and-params.js
"use strict";
/*
 * Copyright (c) "Neo4j"
 * Neo4j Sweden AB [http://neo4j.com]
 *
 * This file is part of Neo4j.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const graphql_relay_1 = require("graphql-relay");
// eslint-disable-next-line import/no-cycle
const create_projection_and_params_1 = __importDefault(require("../create-projection-and-params"));
const create_relationship_property_element_1 = __importDefault(require("../projection/elements/create-relationship-property-element"));
const create_connection_where_and_params_1 = __importDefault(require("../where/create-connection-where-and-params"));
const create_auth_and_params_1 = __importDefault(require("../create-auth-and-params"));
const constants_1 = require("../../constants");
const pagination_1 = require("../../schema/pagination");
const filter_interface_nodes_1 = __importDefault(require("../../utils/filter-interface-nodes"));
function createConnectionAndParams({ resolveTree, field, context, nodeVariable, parameterPrefix, }) {
    var _a, _b;
    let globalParams = {};
    let nestedConnectionFieldParams;
    const subquery = ["CALL {", `WITH ${nodeVariable}`];
    const sortInput = resolveTree.args.sort;
    const afterInput = resolveTree.args.after;
    const firstInput = resolveTree.args.first;
    const whereInput = resolveTree.args.where;
    const relationshipVariable = `${nodeVariable}_${field.relationship.type.toLowerCase()}_relationship`;
    const relationship = context.neoSchema.relationships.find((r) => r.name === field.relationshipTypeName);
    const inStr = field.relationship.direction === "IN" ? "<-" : "-";
    const relTypeStr = `[${relationshipVariable}:${field.relationship.type}]`;
    const outStr = field.relationship.direction === "OUT" ? "->" : "-";
    let relationshipProperties = [];
    let node;
    const connection = resolveTree.fieldsByTypeName[field.typeMeta.name];
    if (connection.edges) {
        const relationshipFieldsByTypeName = connection.edges.fieldsByTypeName[field.relationshipTypeName];
        relationshipProperties = Object.values(relationshipFieldsByTypeName).filter((v) => v.name !== "node");
        node = Object.values(relationshipFieldsByTypeName).find((v) => v.name === "node");
    }
    const elementsToCollect = [];
    if (relationshipProperties.length) {
        const relationshipPropertyEntries = relationshipProperties
            .filter((p) => p.name !== "cursor")
            .map((v) => create_relationship_property_element_1.default({ resolveTree: v, relationship, relationshipVariable }));
        elementsToCollect.push(relationshipPropertyEntries.join(", "));
    }
    if (field.relationship.union || field.relationship.interface) {
        const relatedNodes = field.relationship.union
            ? context.neoSchema.nodes.filter((n) => { var _a, _b; return (_b = (_a = field.relationship.union) === null || _a === void 0 ? void 0 : _a.nodes) === null || _b === void 0 ? void 0 : _b.includes(n.name); })
            : context.neoSchema.nodes.filter((x) => {
                var _a, _b, _c;
                return ((_c = (_b = (_a = field.relationship) === null || _a === void 0 ? void 0 : _a.interface) === null || _b === void 0 ? void 0 : _b.implementations) === null || _c === void 0 ? void 0 : _c.includes(x.name)) &&
                    filter_interface_nodes_1.default({ node: x, whereInput: whereInput === null || whereInput === void 0 ? void 0 : whereInput.node });
            });
        const subqueries = [];
        relatedNodes.forEach((n) => {
            var _a, _b, _c, _d;
            if (!whereInput ||
                Object.prototype.hasOwnProperty.call(whereInput, n.name) ||
                (field.relationship.interface &&
                    !((_b = (_a = field.relationship.interface) === null || _a === void 0 ? void 0 : _a.implementations) === null || _b === void 0 ? void 0 : _b.some((i) => Object.prototype.hasOwnProperty.call(resolveTree.args.where, i))))) {
                const labels = n.getLabelString(context);
                const relatedNodeVariable = `${nodeVariable}_${n.name}`;
                const nodeOutStr = `(${relatedNodeVariable}${labels})`;
                const unionInterfaceSubquery = [];
                const subqueryElementsToCollect = [...elementsToCollect];
                const nestedSubqueries = [];
                if (node) {
                    const nodeFieldsByTypeName = {
                        [n.name]: {
                            ...node === null || node === void 0 ? void 0 : node.fieldsByTypeName[n.name],
                            ...node === null || node === void 0 ? void 0 : node.fieldsByTypeName[field.relationship.typeMeta.name],
                        },
                    };
                    const nodeProjectionAndParams = create_projection_and_params_1.default({
                        fieldsByTypeName: nodeFieldsByTypeName,
                        node: n,
                        context,
                        varName: relatedNodeVariable,
                        literalElements: true,
                        resolveType: true,
                    });
                    const [nodeProjection, nodeProjectionParams] = nodeProjectionAndParams;
                    subqueryElementsToCollect.push(`node: ${nodeProjection}`);
                    globalParams = {
                        ...globalParams,
                        ...nodeProjectionParams,
                    };
                    if ((_d = (_c = nodeProjectionAndParams[2]) === null || _c === void 0 ? void 0 : _c.connectionFields) === null || _d === void 0 ? void 0 : _d.length) {
                        nodeProjectionAndParams[2].connectionFields.forEach((connectionResolveTree) => {
                            const connectionField = n.connectionFields.find((x) => x.fieldName === connectionResolveTree.name);
                            const nestedConnection = createConnectionAndParams({
                                resolveTree: connectionResolveTree,
                                field: connectionField,
                                context,
                                nodeVariable: relatedNodeVariable,
                                parameterPrefix: `${parameterPrefix ? `${parameterPrefix}.` : `${nodeVariable}_`}${resolveTree.alias}.edges.node`,
                            });
                            nestedSubqueries.push(nestedConnection[0]);
                            globalParams = {
                                ...globalParams,
                                ...Object.entries(nestedConnection[1]).reduce((res, [k, v]) => {
                                    if (k !== `${relatedNodeVariable}_${connectionResolveTree.alias}`) {
                                        res[k] = v;
                                    }
                                    return res;
                                }, {}),
                            };
                            if (nestedConnection[1][`${relatedNodeVariable}_${connectionResolveTree.alias}`]) {
                                if (!nestedConnectionFieldParams)
                                    nestedConnectionFieldParams = {};
                                nestedConnectionFieldParams = {
                                    ...nestedConnectionFieldParams,
                                    ...{
                                        [connectionResolveTree.alias]: nestedConnection[1][`${relatedNodeVariable}_${connectionResolveTree.alias}`],
                                    },
                                };
                            }
                        });
                    }
                }
                else {
                    // This ensures that totalCount calculation is accurate if edges not asked for
                    subqueryElementsToCollect.push(`node: { __resolveType: "${n.name}" }`);
                }
                unionInterfaceSubquery.push(`WITH ${nodeVariable}`);
                unionInterfaceSubquery.push(`MATCH (${nodeVariable})${inStr}${relTypeStr}${outStr}${nodeOutStr}`);
                const allowAndParams = create_auth_and_params_1.default({
                    operations: "READ",
                    entity: n,
                    context,
                    allow: {
                        parentNode: n,
                        varName: relatedNodeVariable,
                    },
                });
                if (allowAndParams[0]) {
                    globalParams = { ...globalParams, ...allowAndParams[1] };
                    unionInterfaceSubquery.push(`CALL apoc.util.validate(NOT(${allowAndParams[0]}), "${constants_1.AUTH_FORBIDDEN_ERROR}", [0])`);
                }
                const whereStrs = [];
                const unionInterfaceWhere = field.relationship.union ? (whereInput || {})[n.name] : whereInput || {};
                if (unionInterfaceWhere) {
                    const where = create_connection_where_and_params_1.default({
                        whereInput: unionInterfaceWhere,
                        node: n,
                        nodeVariable: relatedNodeVariable,
                        relationship,
                        relationshipVariable,
                        context,
                        parameterPrefix: `${parameterPrefix ? `${parameterPrefix}.` : `${nodeVariable}_`}${resolveTree.alias}.args.where${field.relationship.union ? `.${n.name}` : ""}`,
                    });
                    const [whereClause] = where;
                    if (whereClause) {
                        if (whereClause) {
                            whereStrs.push(whereClause);
                        }
                    }
                }
                const whereAuth = create_auth_and_params_1.default({
                    operations: "READ",
                    entity: n,
                    context,
                    where: { varName: relatedNodeVariable, node: n },
                });
                if (whereAuth[0]) {
                    whereStrs.push(whereAuth[0]);
                    globalParams = { ...globalParams, ...whereAuth[1] };
                }
                if (whereStrs.length) {
                    unionInterfaceSubquery.push(`WHERE ${whereStrs.join(" AND ")}`);
                }
                if (nestedSubqueries.length) {
                    unionInterfaceSubquery.push(nestedSubqueries.join("\n"));
                }
                unionInterfaceSubquery.push(`WITH { ${subqueryElementsToCollect.join(", ")} } AS edge`);
                unionInterfaceSubquery.push("RETURN edge");
                subqueries.push(unionInterfaceSubquery.join("\n"));
            }
        });
        const subqueryCypher = ["CALL {", subqueries.join("\nUNION\n"), "}"];
        if (sortInput && sortInput.length) {
            const sort = sortInput.map((s) => Object.entries(s.edge || []).map(([f, direction]) => `edge.${f} ${direction}`).join(", "));
            subqueryCypher.push(`WITH edge ORDER BY ${sort.join(", ")}`);
        }
        const withValues = [];
        if (!firstInput && !afterInput) {
            if (connection.edges || connection.pageInfo) {
                withValues.push("collect(edge) as edges");
            }
            withValues.push("count(edge) as totalCount");
            subqueryCypher.push(`WITH ${withValues.join(", ")}`);
        }
        else {
            const offsetLimitStr = pagination_1.createOffsetLimitStr({
                offset: typeof afterInput === "string" ? graphql_relay_1.cursorToOffset(afterInput) + 1 : undefined,
                limit: firstInput,
            });
            subqueryCypher.push("WITH collect(edge) AS allEdges");
            subqueryCypher.push(`WITH allEdges, size(allEdges) as totalCount, allEdges${offsetLimitStr} AS edges`);
        }
        subquery.push(subqueryCypher.join("\n"));
    }
    else {
        const relatedNodeVariable = `${nodeVariable}_${field.relationship.typeMeta.name.toLowerCase()}`;
        const relatedNode = context.neoSchema.nodes.find((x) => x.name === field.relationship.typeMeta.name);
        const labels = relatedNode.getLabelString(context);
        const nodeOutStr = `(${relatedNodeVariable}${labels})`;
        subquery.push(`MATCH (${nodeVariable})${inStr}${relTypeStr}${outStr}${nodeOutStr}`);
        const whereStrs = [];
        if (whereInput) {
            const where = create_connection_where_and_params_1.default({
                whereInput,
                node: relatedNode,
                nodeVariable: relatedNodeVariable,
                relationship,
                relationshipVariable,
                context,
                parameterPrefix: `${parameterPrefix ? `${parameterPrefix}.` : `${nodeVariable}_`}${resolveTree.alias}.args.where`,
            });
            const [whereClause] = where;
            if (whereClause) {
                whereStrs.push(`${whereClause}`);
            }
        }
        const whereAuth = create_auth_and_params_1.default({
            operations: "READ",
            entity: relatedNode,
            context,
            where: { varName: relatedNodeVariable, node: relatedNode },
        });
        if (whereAuth[0]) {
            whereStrs.push(whereAuth[0]);
            globalParams = { ...globalParams, ...whereAuth[1] };
        }
        if (whereStrs.length) {
            subquery.push(`WHERE ${whereStrs.join(" AND ")}`);
        }
        const allowAndParams = create_auth_and_params_1.default({
            operations: "READ",
            entity: relatedNode,
            context,
            allow: {
                parentNode: relatedNode,
                varName: relatedNodeVariable,
            },
        });
        if (allowAndParams[0]) {
            globalParams = { ...globalParams, ...allowAndParams[1] };
            subquery.push(`CALL apoc.util.validate(NOT(${allowAndParams[0]}), "${constants_1.AUTH_FORBIDDEN_ERROR}", [0])`);
        }
        if (sortInput && sortInput.length) {
            const sort = sortInput.map((s) => [
                ...Object.entries(s.edge || []).map(([f, direction]) => `${relationshipVariable}.${f} ${direction}`),
                ...Object.entries(s.node || []).map(([f, direction]) => `${relatedNodeVariable}.${f} ${direction}`),
            ].join(", "));
            subquery.push(`WITH ${relationshipVariable}, ${relatedNodeVariable}`);
            subquery.push(`ORDER BY ${sort.join(", ")}`);
        }
        const nestedSubqueries = [];
        if (node) {
            const nodeProjectionAndParams = create_projection_and_params_1.default({
                fieldsByTypeName: node === null || node === void 0 ? void 0 : node.fieldsByTypeName,
                node: relatedNode,
                context,
                varName: relatedNodeVariable,
                literalElements: true,
            });
            const [nodeProjection, nodeProjectionParams, projectionMeta] = nodeProjectionAndParams;
            elementsToCollect.push(`node: ${nodeProjection}`);
            globalParams = { ...globalParams, ...nodeProjectionParams };
            if ((_a = projectionMeta === null || projectionMeta === void 0 ? void 0 : projectionMeta.authValidateStrs) === null || _a === void 0 ? void 0 : _a.length) {
                subquery.push(`CALL apoc.util.validate(NOT(${projectionMeta.authValidateStrs.join(" AND ")}), "${constants_1.AUTH_FORBIDDEN_ERROR}", [0])`);
            }
            if ((_b = projectionMeta === null || projectionMeta === void 0 ? void 0 : projectionMeta.connectionFields) === null || _b === void 0 ? void 0 : _b.length) {
                projectionMeta.connectionFields.forEach((connectionResolveTree) => {
                    const connectionField = relatedNode.connectionFields.find((x) => x.fieldName === connectionResolveTree.name);
                    const nestedConnection = createConnectionAndParams({
                        resolveTree: connectionResolveTree,
                        field: connectionField,
                        context,
                        nodeVariable: relatedNodeVariable,
                        parameterPrefix: `${parameterPrefix ? `${parameterPrefix}.` : `${nodeVariable}_`}${resolveTree.alias}.edges.node`,
                    });
                    nestedSubqueries.push(nestedConnection[0]);
                    globalParams = {
                        ...globalParams,
                        ...Object.entries(nestedConnection[1]).reduce((res, [k, v]) => {
                            if (k !== `${relatedNodeVariable}_${connectionResolveTree.alias}`) {
                                res[k] = v;
                            }
                            return res;
                        }, {}),
                    };
                    if (nestedConnection[1][`${relatedNodeVariable}_${connectionResolveTree.alias}`]) {
                        if (!nestedConnectionFieldParams)
                            nestedConnectionFieldParams = {};
                        nestedConnectionFieldParams = {
                            ...nestedConnectionFieldParams,
                            ...{
                                [connectionResolveTree.alias]: nestedConnection[1][`${relatedNodeVariable}_${connectionResolveTree.alias}`],
                            },
                        };
                    }
                });
            }
        }
        if (nestedSubqueries.length)
            subquery.push(nestedSubqueries.join("\n"));
        subquery.push(`WITH collect({ ${elementsToCollect.join(", ")} }) AS edges`);
    }
    const returnValues = [];
    if (!firstInput && !afterInput) {
        if (connection.edges || connection.pageInfo) {
            returnValues.push("edges: edges");
        }
        returnValues.push(`totalCount: ${field.relationship.union ? "totalCount" : "size(edges)"}`);
        subquery.push(`RETURN { ${returnValues.join(", ")} } AS ${resolveTree.alias}`);
    }
    else {
        const offsetLimitStr = pagination_1.createOffsetLimitStr({
            offset: typeof afterInput === "string" ? graphql_relay_1.cursorToOffset(afterInput) + 1 : undefined,
            limit: firstInput,
        });
        subquery.push(`WITH size(edges) AS totalCount, edges${offsetLimitStr} AS limitedSelection`);
        subquery.push(`RETURN { edges: limitedSelection, totalCount: totalCount } AS ${resolveTree.alias}`);
    }
    subquery.push("}");
    const params = {
        ...globalParams,
        ...((whereInput || nestedConnectionFieldParams) && {
            [`${nodeVariable}_${resolveTree.alias}`]: {
                ...(whereInput && { args: { where: whereInput } }),
                ...(nestedConnectionFieldParams && { edges: { node: { ...nestedConnectionFieldParams } } }),
            },
        }),
    };
    return [subquery.join("\n"), params];
}
exports.default = createConnectionAndParams;
