// @neo4j/graphql/dist/translate/where/create-filter.js
"use strict";
/*
 * Copyright (c) "Neo4j"
 * Neo4j Sweden AB [http://neo4j.com]
 *
 * This file is part of Neo4j.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.Operator = void 0;
var Operator;
(function (Operator) {
    Operator["INCLUDES"] = "IN";
    Operator["IN"] = "IN";
    Operator["MATCHES"] = "=~";
    Operator["CONTAINS"] = "CONTAINS";
    Operator["STARTS_WITH"] = "STARTS WITH";
    Operator["ENDS_WITH"] = "ENDS WITH";
    Operator["LT"] = "<";
    Operator["GT"] = ">";
    Operator["GTE"] = ">=";
    Operator["LTE"] = "<=";
    Operator["DISTANCE"] = "=";
})(Operator = exports.Operator || (exports.Operator = {}));
function createFilter({ left, operator, right, not, }) {
    if (!Operator[operator]) {
        throw new Error(`Invalid filter operator ${operator}`);
    }
    if (not && ["MATCHES", "LT", "GT", "GTE", "LTE", "DISTANCE"].includes(operator)) {
        throw new Error(`Invalid filter operator NOT_${operator}`);
    }
    let filter = `${left} ${Operator[operator]} ${right}`;
    if (not)
        filter = `(NOT ${filter})`;
    return filter;
}
exports.default = createFilter;
